### ISSUE TITLE
Remove aria-pressed from expand/collapse buttons

### ISSUE THREAD

**Issue by LJWatson** [stevefaulkner/HTML5accessibility#26]
Would remove aria-pressed from the expand/collapse buttons. The attribute makes it a toggle button, in which case the label of the button shouldn't change.

From ARIA 1.1 APG 2.6 button [1]:

"Toggle button: A two-state button that can be either off (not pressed) or on (pressed). To tell assistive technologies that a button is a toggle button, specify a value for the attribute aria-pressed… Important: the label on a toggle button should never change when the state changes."

[1] https://www.w3.org/TR/2016/WD-wai-aria-practices-1.1-20160317/#button

**Comment by dstorey** [comment]
Thanks @LjWatson. This is the main area of the design I wanted help with getting right, as it is a fairly complex table, with sub rows (for the criteria that goes into an element/property passing/failing/partial). Cynthia a tree grid or such, that I need to look into.

Does it give enough context with just the label changing and I should just have no aria? I could also keep the attribute and keep the label the same, but I'm not sure what that would be. The other one I was considering was aria-expanded, which may make more sense than aria-pressed. If that is the best option, I guess it should also not change it's text? I guess maybe something like "show test criteria" (although that is wordy, and maybe even more weird to be announced for every top level row)

Does what is read out make sense for that cell, as the accessible name becomes "expand [element name]"? If that doesn't make sense I'm not sure a way around it, as both are in the DOM in the table cell. I originally put it after the text content and floated left, but that made focus order different to visual order (link to element spec, then expand button), and Cynthia rightly mentioned that in normal expandable tree things, the arrow is always to the left anyway.

**Comment by LjWatson** [comment]
@dstorey
I've looked more at this and have some thoughts...

The accessible name for each button is either "Expand" or "Collapse". The element name is not associated with the button, despite the fact they're both inside the same ```<td>```.

The element name isn't sufficiently useful as link text. Until I looked at the code I assumed those links pointed to the spec definition for the relevant element, and not to the test files as they do.

Would a link to the relevant spec definition also be useful?

The simplest solution might be to include the element name as part of the button label, and to use aria-expanded to indicate the current state of the information. For example:
```
<button aria-expanded="false">article details</button>
```
If that would be too much information to be comfortable visually, then simplifying the visible label to "Details" and providing the longer version using ARIA might be an idea. For example:
```
<button aria-expanded="false" aria-label="article details">Details</button>
```

**Comment by dstorey** [comment]
@LjWatson comments:

* The button text is visually hidden and replaced by a plus or minus icon (depending on expand/collapse state), so it can contain which ever text you think is most appropriate. The only limit I have on the button text is that they're programatically created to make it easier to maintain the file by hand and stop people making mistakes when editing. I could probably just pull the text content from the links, or perhaps the full  TD.

For the link text, were you expecting it to be the spec, mostly because you're used to editing specs where that is the link text? There isn't any thing in the text to suggest it is the spec or the test. I'd prefer to not add a spec link as it is double the maintenance, more UI to scan through, and the spec is linked from the test pages. I'd argue that for this test the spec for the element is the least important spec. The HTML and Core AAM specs, plus perhaps the name/description specs are more relevant to how accessible the elements are. All of those are linked from the test pages too.

Would the links be more clear if the accessible description / help text for each one mentioned where they go? For this I'd probably use the title attribute as it is just as much an issue for sighted users as screen reader users. I was thinking something like:

````HTML
<a href="#" title="test file for article">article</a>
````

To make that easier to maintain I could probably do the same as the buttons and inject via JS by pulling in the anchor element text content.

**Comment by dstorey** [comment]
The other option I think it to add role="treegrid" to the table, and then specify aria-expanded on the rows, and aria-level="1" on the parent rows (e.g. article), and aria-level="2" on the sub-criteria rows. If I do this, I guess the rows get the click event and I should make a fake button for visual users that is hidden to screen reader users. Although this might get in the way of being able to click links inside the table cells. I'm still looking for a good treegrid example. I assume if I use tree grid, I don't have to add the aria roles to the rows and cells, as they match a normal data table's default roles, but that is not 100% clear.  This is way more complex though, as probably need to add a bunch of keyboard handling.

**Comment by stevefaulkner** [comment]
On 25 April 2016 at 01:59, David Storey <notifications@github.com> wrote:

> There isn't any thing in the text to suggest it is the spec or the test.
> I'd prefer to not add a spec link as it is double the maintenance, more UI
> to scan through, and the spec is linked from the test pages. I'd argue that
> for this test the spec for the element is the least important spec.


+1

--

Regards

SteveF
Current Standards Work @W3C
<http://www.paciellogroup.com/blog/2015/03/current-standards-work-at-w3c/>

**Comment by LjWatson** [comment]
@dstorey
"For the link text, were you expecting it to be the spec, mostly because you're used to editing specs where that is the link text?"

Probably :) The question about linking to the spec was an aside though. The thing I wanted to surface was that the link text doesn't clearly indicate where the link leads. "Element" isn't enough information IMO, so suggest the link text should be "Element test" instead.

"For this I'd probably use the title attribute as it is just as much an issue for sighted users as screen reader users."

It would definitely be helpful for the link text to be visible/available to all, but the title attribute is fraught with problems (obscures content underneath for magnifier users, not supported on mobile etc.). More info:
https://www.paciellogroup.com/blog/2013/01/using-the-html-title-attribute-updated/ 

"The other option I think it to add role="treegrid" to the table, and then specify aria-expanded on the rows, and aria-level="1" on the parent rows (e.g. article), and aria-level="2" on the sub-criteria rows."

This would complicate things in terms of implementation and interaction I think. Plus AFAIK the ARIA roles for trees are not as widely supported as we'd need them to be.

The pattern being used now is simple (which is good). If we can just figure out a neat way to improve the button and link labelling, we'll be sorted.

**Comment by sandrina-p** [comment]
Hey guys! Old question but here I am within a similar scenario.

On a personal project, I have Tabs toggled by different buttons that have `aria-expanded`. Now I don't know if I should also add `aria-pressed` or not...
First though I think is redundant, but then I read what you've said and I'll keep both.

You can see on [my website](http://sandrina-p), on "Projects" link, I have a "carousel / tabs" where I'm not sure if using only `aria-expanded` for each Project button is enough. What do you think would fit the best? 🤔
